### PR TITLE
Adding timeformat field to Schema

### PR DESCRIPTION
**Pull request by kishoreg**
Currently, Schema does not allow us to represent the time format and assumes its always sinceEpoch format. 

This PR add timeFormat field to the schema. Now one can specify the format is in either in SIMPLE_DATE_FORMAT or EPOCH format. The default is EPOCH to maintain backward compatibility.

Unfortunately,this is not forward compatible i.e the if anyone uploads schema with timeFormat, then the old code will fail to parse. We should first roll out this version before we upload schema with timeFormat field

Have a couple of TODO's which is beyond the scope of this PR. 
- conversion between sdf <-> sdf/epoch is not supported. Will file a jira for this.
- I did not add additional builder methods to SchemaBuilder, we already have 13 methods to addTime. Will create another issue to reduce it to 2.

### CODE REVIEW

File: pinot-common/src/main/java/com/linkedin/pinot/common/data/TimeGranularitySpec.java [hunk 1]
```diff
@@ -35,14 +37,22 @@
...
   private int _timeUnitSize = DEFAULT_TIME_UNIT_SIZE;
+  /*
+   * Can be either EPOCH (default) or SIMPLE_DATE_FORMAT:format e.g SIMPLE_DATE_FORMAT:yyyyMMdd 
+   */
+  private String _timeFormat = TimeFormat.EPOCH.toString();
   private String _name;
-
+  public enum TimeFormat {
+    EPOCH, //default
+    SIMPLE_DATE_FORMAT
```
Comment (mcvsubbu) [inline pinot-common/src/main/java/com/linkedin/pinot/common/data/TimeGranularitySpec.java]: Please add a block of comments with these two enums with examples of what values may be contained in the time field. Also, be sure to mention that the time field is a 'long', no matter what the time format is.

File: pinot-common/src/main/java/com/linkedin/pinot/common/data/TimeGranularitySpec.java [hunk 2]
```diff
@@ -55,7 +65,18 @@ public TimeGranularitySpec(@Nonnull DataType dataType, @Nonnull TimeUnit timeTyp
...
+  
+  public TimeGranularitySpec(@Nonnull DataType dataType, @Nonnull TimeUnit timeType, @Nonnull String timeFormat,
+      @Nonnull String name) {
+    Preconditions.checkNotNull(timeType);
+    Preconditions.checkNotNull(name);
 
+    _dataType = dataType.getStoredType();
+    _timeType = timeType;
+    _name = name;
+    _timeFormat = timeFormat;
```
Comment (mcvsubbu) [inline pinot-common/src/main/java/com/linkedin/pinot/common/data/TimeGranularitySpec.java]: Should we not do some validation on the time format here? (should start with SImpleDateFormat, and should have a valid string that is parsable by DateFormat?

File: pinot-common/src/main/java/com/linkedin/pinot/common/data/TimeGranularitySpec.java [hunk 3]
```diff
@@ -55,7 +65,18 @@ public TimeGranularitySpec(@Nonnull DataType dataType, @Nonnull TimeUnit timeTyp
     _timeType = timeType;
     _name = name;
   }
+  
+  public TimeGranularitySpec(@Nonnull DataType dataType, @Nonnull TimeUnit timeType, @Nonnull String timeFormat,
```
Comment (mcvsubbu) [inline pinot-common/src/main/java/com/linkedin/pinot/common/data/TimeGranularitySpec.java]: Can you please add a javadoc and provide examples on what 'timeFormat' String values can be? thanks

File: pinot-common/src/main/java/com/linkedin/pinot/common/utils/time/DefaultTimeConverter.java [hunk 4]
```diff
@@ -16,15 +16,27 @@
...
 
   TimeGranularitySpec incoming;
   TimeGranularitySpec outgoing;
+  private boolean conversionSupported;
 
   public void init(TimeGranularitySpec incoming, TimeGranularitySpec outgoing) {
     this.incoming = incoming;
     this.outgoing = outgoing;
+    conversionSupported = false;
+    if (TimeFormat.EPOCH.toString().equals(incoming.getTimeFormat())
```
Comment (kishoreg) [inline pinot-common/src/main/java/com/linkedin/pinot/common/utils/time/DefaultTimeConverter.java]: we already support converting between epoch formats e.g millis to hoursSinceEpoch, millis to daysSinceEpoch etc

File: pinot-common/src/main/java/com/linkedin/pinot/common/utils/time/DefaultTimeConverter.java [hunk 5]
```diff
@@ -16,15 +16,27 @@
...
+        && TimeFormat.EPOCH.toString().equals(outgoing.getTimeFormat())) {
+      conversionSupported = true;
+    }
+    if (conversionSupported) {
```
Comment (kishoreg) [inline pinot-common/src/main/java/com/linkedin/pinot/common/utils/time/DefaultTimeConverter.java]: That's right. will fix it.

File: pinot-common/src/main/java/com/linkedin/pinot/common/data/TimeGranularitySpec.java (hunk 3 above)
Comment (kishoreg) [inline pinot-common/src/main/java/com/linkedin/pinot/common/data/TimeGranularitySpec.java]: will do

File: pinot-common/src/main/java/com/linkedin/pinot/common/data/TimeGranularitySpec.java (hunk 1 above)
Comment (kishoreg) [inline pinot-common/src/main/java/com/linkedin/pinot/common/data/TimeGranularitySpec.java]: I was expecting to support string format as well. But you might be right, we might be assuming long in our code.

File: pinot-common/src/main/java/com/linkedin/pinot/common/data/TimeGranularitySpec.java (hunk 2 above)
Comment (kishoreg) [inline pinot-common/src/main/java/com/linkedin/pinot/common/data/TimeGranularitySpec.java]: will do.

**Comment by fx19880617** [comment]
waiting for this diff for one year long:)

File: pinot-common/src/main/java/com/linkedin/pinot/common/data/TimeGranularitySpec.java (hunk 1 above)
Comment (kishoreg) [inline pinot-common/src/main/java/com/linkedin/pinot/common/data/TimeGranularitySpec.java]: SDF supports ISO format right?

Comment (mcvsubbu) [inline pinot-common/src/main/java/com/linkedin/pinot/common/data/TimeGranularitySpec.java]: Assumption of long time field is made in the controller as well, for retention calculation, segment metadata, etc.

File: pinot-common/src/main/java/com/linkedin/pinot/common/data/TimeGranularitySpec.java [hunk 6]
```diff
@@ -17,6 +17,7 @@
 
 import com.google.common.base.Preconditions;
 import com.linkedin.pinot.common.data.FieldSpec.DataType;
+import com.linkedin.pinot.common.data.TimeGranularitySpec.TimeFormat;
```
Comment (mcvsubbu) [inline pinot-common/src/main/java/com/linkedin/pinot/common/data/TimeGranularitySpec.java]: LGTM

**Pull request merged**